src/cart/src/cart/crt_hg.c

```diff
@@ -1369,24 +1341,43 @@ crt_hg_progress(struct crt_hg_context *hg_ctx, int64_t timeout)
 	else
 		hg_timeout = timeout / 1000;
 
-	rc = crt_hg_trigger(hg_ctx);
-	if (rc != 0)
-		return rc;
+	do {
+		hg_return_t     hg_ret = HG_SUCCESS;
+		int             rc = 0;
+		unsigned int count = 0;
 
-	/** progress RPC execution */
-	hg_ret = HG_Progress(hg_context, hg_timeout);
-	if (hg_ret == HG_TIMEOUT)
-		D_GOTO(out, rc = -DER_TIMEDOUT);
-	if (hg_ret != HG_SUCCESS) {
-		D_ERROR("HG_Progress failed, hg_ret: %d.\n", hg_ret);
-		D_GOTO(out, rc = -DER_HG);
-	}
+		/** progress RPC execution */
+		hg_ret = HG_Progress(hg_context, hg_timeout);
+		if (hg_ret == HG_TIMEOUT) {
+			rc = -DER_TIMEDOUT;
+		} else if (hg_ret != HG_SUCCESS) {
+			D_ERROR("HG_Progress failed, hg_ret: %d.\n", hg_ret);
+			return -DER_HG;
+		}
 
-	/* some RPCs have progressed, call Trigger again */
-	rc = crt_hg_trigger(hg_ctx);
+		/** some RPCs have progressed, call Trigger */
+		hg_ret = HG_Trigger(hg_context, 0, total, &count);
```

Comment:
if HG_Progress got TIMEOUT, do we still need to call HG_Trigger()? seems need not, but probably fine to keep the code as this patch firstly.

Reply:
I actually thought about this when I changed this code and first thought that this is indeed not required. Then i changed my mind since HG_Progress might have progressed 257 things and HG_Trigger() still needs to be called the 2nd time to progress the last item that hasn't been triggered yet. In any case, it should be pretty harmless.